File: src/main/java/org/manapart/enderports/TeleporterNetwork.java
package org.manapart.enderports;

import net.minecraft.nbt.CompoundNBT;
import net.minecraft.nbt.INBT;
import net.minecraft.nbt.ListNBT;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.ServerWorld;
import net.minecraft.world.World;
import net.minecraft.world.storage.WorldSavedData;
import net.minecraftforge.common.util.Constants;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.function.Supplier;

import static org.manapart.enderports.EnderPorts.MODID;

public class TeleporterNetwork extends WorldSavedData {
    public static final String DATA_NAME = MODID + "_TeleporterSaveData";
    private HashMap<String, ArrayList<BlockPos>> network = new HashMap<>();
    private World world;

    public TeleporterNetwork(World world) {
        super(DATA_NAME);
        this.world = world;
    }

    @Override
    public void read(CompoundNBT nbt) {
        ListNBT nodes = nbt.getList("nodes", Constants.NBT.TAG_COMPOUND);
        for (INBT nodeI : nodes) {
            CompoundNBT node = (CompoundNBT) nodeI;
            String key = node.getString("key");
            double x = node.getDouble("x");
            double y = node.getDouble("y");
            double z = node.getDouble("z");
            BlockPos pos = new BlockPos(x, y, z);
            addTeleporter(key, pos);
        }
    }

    @Override
    public CompoundNBT write(CompoundNBT compound) {
        CompoundNBT cnbt = new CompoundNBT();
        ListNBT nodes = new ListNBT();

        for (String key : network.keySet()) {
            for (BlockPos value : network.get(key)) {
                CompoundNBT node = new CompoundNBT();
                node.putString("key", key);
                node.putDouble("x", value.getX());
                node.putDouble("y", value.getY());
                node.putDouble("z", value.getZ());
                nodes.add(node);
            }
        }

        cnbt.put("nodes", nodes);
        return cnbt;
    }

    public void addTeleporter(BlockPos pos) {
        String beneathBlockName = getKey(pos);
        addTeleporter(beneathBlockName, pos);
    }


    public void addTeleporter(String beneathBlockName, BlockPos pos) {
//        System.out.println("Adding teleporter with " + beneathBlockName);
        if (!network.containsKey(beneathBlockName)) {
            network.put(beneathBlockName, new ArrayList<>());
        }
        if (!network.get(beneathBlockName).contains(pos)) {
            network.get(beneathBlockName).add(pos);
        }
        markDirty();
    }

    public void removeTeleporter(BlockPos pos) {
        String beneathBlockName = getKey(pos);
//        System.out.println("Removing teleporter with " + beneathBlockName);
        if (network.containsKey(beneathBlockName) && network.get(beneathBlockName).contains(pos)) {
            network.get(beneathBlockName).remove(pos);
        }
        markDirty();
    }

    public BlockPos getNextTeleporter(BlockPos pos) {
        return getNextTeleporterWithRetry(pos, true);
    }

    private BlockPos getNextTeleporterWithRetry(BlockPos pos, boolean retry) {
        String beneathBlockName = getKey(pos);
//        System.out.println("Next teleporter with " + beneathBlockName);
        if (network.containsKey(beneathBlockName) && network.get(beneathBlockName).size() > 0) {
            ArrayList<BlockPos> positions = network.get(beneathBlockName);
            int index = positions.indexOf(pos) + 1;
            if (index >= positions.size()) index = 0;
            BlockPos nextPos = positions.get(index);
            if (isTeleporter(nextPos)) {
                if (getKey(nextPos).equals(beneathBlockName)) {
                    return nextPos;
                }
            } else {
                removeTeleporter(nextPos);
            }
        }
        if (retry){
            reBalance();
            return getNextTeleporterWithRetry(pos, false);
        }
        return pos;
    }

    public void reBalance() {
        System.out.println("Rebalancing teleporter network.");
        HashMap<String, ArrayList<BlockPos>> staleTeleporters = new HashMap<>();
        for (String beneathBlockName : network.keySet()) {
            for (BlockPos pos : network.get(beneathBlockName)) {
                if (!getKey(pos).equals(beneathBlockName)) {
                    if (!staleTeleporters.containsKey(beneathBlockName)) {
                        staleTeleporters.put(beneathBlockName, new ArrayList<>());
                    }
                    staleTeleporters.get(beneathBlockName).add(pos);
                }
            }
        }

        for (String beneathBlockName : staleTeleporters.keySet()) {
            for (BlockPos pos : staleTeleporters.get(beneathBlockName)) {
                network.get(beneathBlockName).remove(pos);
                addTeleporter(pos);
            }
        }
    }

    private String getKey(BlockPos pos) {
        ResourceLocation registryName = world.getBlockState(pos.down()).getBlock().getRegistryName();
        if (registryName != null) {
            return registryName.toString();
        }
        return "";
    }

    private boolean isTeleporter(BlockPos pos) {
        ResourceLocation blockRegistryKey = world.getBlockState(pos).getBlock().getRegistryName();
        String key = "";
        if (blockRegistryKey != null) {
            key = blockRegistryKey.toString();
        }
        return key.equals(EnderPorts.teleporter.getRegistryName().toString());
    }

    public static TeleporterNetwork getNetwork(ServerWorld world) {
        return world.getSavedData().getOrCreate(new NetworkSupplier(world), DATA_NAME);
    }

    private static class NetworkSupplier implements Supplier<TeleporterNetwork> {
        private ServerWorld world;

        public NetworkSupplier(ServerWorld world) {
            this.world = world;
        }

        @Override
        public TeleporterNetwork get() {
            return new TeleporterNetwork(world);
        }
    }

}

File: src/main/java/org/manapart/enderports/TeleporterNetwork.kt
package org.manapart.enderports

import net.minecraft.core.BlockPos
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.level.Level
import net.minecraft.world.level.saveddata.SavedData
import java.util.function.Supplier


const val DATA_NAME = MODID + "_TeleporterSaveData"

class TeleporterNetwork(private val world: Level) : SavedData() {
    private val network = mutableMapOf<String, MutableSet<BlockPos>>()
    private var teleporterChain = mapOf<BlockPos, BlockPos>()

    override fun save(cnbt: CompoundTag): CompoundTag {
        val nodes = ListTag()
        for (key in network.keys) {
            for (value in network[key]!!) {
                val node = CompoundTag()
                node.putString("key", key)
                node.putDouble("x", value.x.toDouble())
                node.putDouble("y", value.y.toDouble())
                node.putDouble("z", value.z.toDouble())
                nodes.add(node)
            }
        }
        cnbt.put("nodes", nodes)
        return cnbt
    }

    fun addTeleporter(pos: BlockPos) {
        val beneathBlockName = getKey(pos)
        addTeleporter(beneathBlockName, pos)
    }

    internal fun addTeleporter(beneathBlockName: String, pos: BlockPos) {
        network.putIfAbsent(beneathBlockName, mutableSetOf())
        network[beneathBlockName]?.add(pos)
        buildTeleporterChain()
        setDirty()
    }

    fun removeTeleporter(pos: BlockPos) {
        val beneathBlockName = getKey(pos)
        if (network.containsKey(beneathBlockName) && network[beneathBlockName]?.contains(pos) == true) {
            network[beneathBlockName]?.remove(pos)
            buildTeleporterChain()
            setDirty()
            reBalance(beneathBlockName)
        }
    }

    fun getNextTeleporter(pos: BlockPos): BlockPos {
        return teleporterChain[pos] ?: pos
    }

    //In the case we teleport someone to a stale location, at least remove it so it doesn't happen again
    fun removeStaleLocation(pos: BlockPos) {
        if (!isTeleporter(pos)) {
            removeTeleporter(pos)
        }
    }

    private fun reBalance(key: String) {
        println("Balancing teleporter network for $key")
        val start = System.currentTimeMillis()

        if (network[key] != null) {
            val validSpots = network[key]?.filter { isTeleporter(it) }?.toMutableSet() ?: mutableSetOf()
            if (validSpots != network[key]) {
                network[key] = validSpots
                buildTeleporterChain()
                setDirty()
            }
        }

        println("Rebalance complete in "+ (System.currentTimeMillis() - start))
    }

    internal fun buildTeleporterChain() {
        val newChain = mutableMapOf<BlockPos, BlockPos>()
        network.values.map { it.toList() }.forEach { chain ->
            (0 until chain.size - 1).forEach { i ->
                newChain[chain[i]] = chain[i + 1]
            }
            newChain[chain.last()] = chain.first()
        }
        teleporterChain = newChain
    }

    private fun getKey(pos: BlockPos): String {
        return try {
            val state = world.getBlockState(pos.below())
            state.block.descriptionId.toString()
        } catch (e: Exception) {
            println("Unable to find teleporter key for $pos")
            ""
        }
    }

    private fun isTeleporter(pos: BlockPos): Boolean {
        val key = world.getBlockState(pos).block.descriptionId.toString()
        return key == ModBlocks.TELEPORTER_BLOCK.descriptionId.toString()
    }

    fun dumpText(): String {
        return network.entries.joinToString("\n") { (blockId, positions) ->
            blockId + "\n\t" + positions.joinToString(",") { "(${it.x},${it.y},${it.z})" }
        }
    }

    internal class NetworkSupplier(private val world: Level) : Supplier<TeleporterNetwork> {
        override fun get(): TeleporterNetwork = TeleporterNetwork(world)
    }
}

fun load(nbt: CompoundTag, world: Level): TeleporterNetwork {
    val network = TeleporterNetwork(world)
    //Constants.NBT.TAG_COMPOUND - not sure where this constant lives now
    nbt.getList("nodes", 10).forEach {
        val node = it as CompoundTag
        val key = node.getString("key")
        val x = node.getDouble("x")
        val y = node.getDouble("y")
        val z = node.getDouble("z")
        val pos = BlockPos(x, y, z)
        network.addTeleporter(key, pos)
    }
    network.buildTeleporterChain()
//    println("Teleport Network Loaded")
//    println(network.dumpText())
    return network
}

fun ServerLevel.getNetwork(): TeleporterNetwork {
    val loadFunction = { nbt: CompoundTag -> load(nbt, this) }
    return dataStorage.computeIfAbsent(loadFunction, TeleporterNetwork.NetworkSupplier(this), DATA_NAME)
}